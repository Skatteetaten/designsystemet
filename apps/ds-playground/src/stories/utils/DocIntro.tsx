import { ReactNode } from 'react';

import { Link } from '@skatteetaten/ds-buttons';
import { Paragraph } from '@skatteetaten/ds-typography';

interface DocIntroProps {
  name: string;
  url: string;
}

export const DocIntro = ({ name, url }: DocIntroProps): ReactNode => {
  return (
    <Paragraph variant={'ingress'}>
      {`Se `}
      <Link href={url}>{name}</Link>
      {` p√• stil og tone for eksempler og
        retningslinjer for denne komponenten.`}
    </Paragraph>
  );
};
