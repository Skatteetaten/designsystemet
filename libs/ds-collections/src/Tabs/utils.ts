export const valueRegex = /^[A-Za-z][\w\-.:]*$/;
