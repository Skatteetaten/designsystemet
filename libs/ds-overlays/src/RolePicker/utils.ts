import { formatter } from '@skatteetaten/ds-core-utils';

export const formatOrganisationNumber = (
  organisationNumber: string
): string => {
  const result = formatter({
    type: 'organisationNumber',
    value: organisationNumber,
  });

  return result.value;
};

export const formatNationalIdentityNumber = (
  nationalIdentityNumber: string
): string => {
  const result = formatter({
    type: 'nationalIdentityNumber',
    value: nationalIdentityNumber,
  });

  return result.value;
};
