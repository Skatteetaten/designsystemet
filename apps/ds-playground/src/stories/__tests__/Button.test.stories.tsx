import { FocusEvent, MouseEvent, useState } from 'react';

import { StoryFn, Meta, StoryObj } from '@storybook/react';
import { expect, fn, userEvent, waitFor, within } from '@storybook/test';

import { Button, buttonVariantArr } from '@skatteetaten/ds-buttons';
import {
  dsI18n,
  getCommonButtonTypeDefault,
} from '@skatteetaten/ds-core-utils';
import { SendSVGpath } from '@skatteetaten/ds-icons';

import { wrapper } from './testUtils/storybook.testing.utils';
import { SystemSVGPaths } from '../utils/icon.systems';

const defaultButtonText = 'Klikk her';

const verifyAttribute =
  (attribute: string, expectedValue: string) =>
  async ({ canvasElement }: { canvasElement: HTMLElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button');
    await expect(button).toBeInTheDocument();
    await expect(button).toHaveAttribute(attribute, expectedValue);
  };

const meta = {
  component: Button,
  title: 'Tester/Button',
  argTypes: {
    // Baseprops
    key: { table: { disable: true } },
    ref: { table: { disable: true } },
    className: { table: { disable: true } },
    id: { table: { disable: true } },
    lang: { table: { disable: true } },
    'data-testid': { table: { disable: true } },
    // Props
    children: { table: { disable: true } },
    hasSpinner: { table: { disable: true } },
    spinnerTitle: { table: { disable: true } },
    svgPath: {
      table: { disable: true },
      options: Object.keys(SystemSVGPaths),
      mapping: SystemSVGPaths,
    },
    variant: {
      table: { disable: true },
      options: [...buttonVariantArr],
      control: 'radio',
    },
    isExternal: { table: { disable: true } },
    // HTML
    accessKey: { table: { disable: true } },
    disabled: { table: { disable: true } },
    href: { table: { disable: true } },
    form: { table: { disable: true } },
    type: { table: { disable: true } },
    // Aria
    ariaDescribedby: { table: { disable: true } },
    ariaCurrent: { table: { disable: true } },
    // Events
    onBlur: { table: { disable: true } },
    onClick: { table: { disable: true } },
    onFocus: { table: { disable: true } },
  },
} satisfies Meta<typeof Button>;
export default meta;
type Story = StoryObj<typeof meta>;

const defaultArgs = {
  children: defaultButtonText,
};

export const WithRef = {
  name: 'With Ref (FA1)',
  args: {
    ...defaultArgs,
    ref: (instance: HTMLButtonElement | null): void => {
      if (instance) {
        instance.id = 'dummyIdForwardedFromRef';
      }
    },
  },
  argTypes: {
    ref: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: verifyAttribute('id', 'dummyIdForwardedFromRef'),
} satisfies Story;

export const WithAttributes = {
  name: 'With Attributes (FA2-5)',
  args: {
    ...defaultArgs,
    id: 'htmlid',
    className: 'dummyClassname',
    lang: 'nb',
    'data-testid': '123ID',
    form: 'formid123',
  },
  argTypes: {
    id: { table: { disable: false } },
    className: { table: { disable: false } },
    lang: { table: { disable: false } },
    'data-testid': { table: { disable: false } },
    form: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: {
      focus: `${wrapper} > button`,
      hover: `${wrapper} > button`,
      click: `${wrapper} > button`,
    },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button');
    await expect(button).toHaveAttribute('id', 'htmlid');
    await expect(button).toHaveClass('dummyClassname');
    await expect(button).toHaveAttribute('lang', 'nb');
    await expect(button).toHaveAttribute('data-testid', '123ID');
    await expect(button).toHaveAttribute('form', 'formid123');
  },
} satisfies Story;

export const Defaults = {
  name: 'Defaults Variant Primary (A1, B1)',
  args: {
    ...defaultArgs,
  },
  argTypes: {
    children: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: {
      focus: `${wrapper} > button`,
      hover: `${wrapper} > button`,
      click: `${wrapper} > button`,
    },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button');
    await expect(button).toBeInTheDocument();
    await expect(button).toHaveAttribute('type', getCommonButtonTypeDefault());
  },
} satisfies Story;

export const VariantSecondary = {
  name: 'Variant Secondary (A1)',
  args: {
    ...defaultArgs,
    variant: 'secondary',
  },
  argTypes: {
    variant: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: {
      focus: `${wrapper} > button`,
      hover: `${wrapper} > button`,
      click: `${wrapper} > button`,
    },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button');
    await expect(button).toBeInTheDocument();
    await expect(button).toHaveAttribute('type', getCommonButtonTypeDefault());
  },
} satisfies Story;

export const VariantTertiary = {
  name: 'Variant Tertiary (A1)',
  args: {
    ...defaultArgs,
    variant: 'tertiary',
  },
  argTypes: {
    variant: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: {
      focus: `${wrapper} > button`,
      hover: `${wrapper} > button`,
      click: `${wrapper} > button`,
    },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button');
    await expect(button).toBeInTheDocument();
    await expect(button).toHaveAttribute('type', getCommonButtonTypeDefault());
  },
} satisfies Story;

export const VariantDanger = {
  name: 'Variant Danger (A1)',
  args: { ...defaultArgs, variant: 'danger' },
  argTypes: {
    variant: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: {
      focus: `${wrapper} > button`,
      hover: `${wrapper} > button`,
      click: `${wrapper} > button`,
    },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button');
    await expect(button).toBeInTheDocument();
    await expect(button).toHaveAttribute('type', getCommonButtonTypeDefault());
  },
} satisfies Story;

export const WithIcon = {
  name: 'With Icon (A3, B3)',
  args: {
    ...defaultArgs,
    svgPath: SendSVGpath,
  },
  argTypes: {
    svgPath: { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button');

    const svg = button.querySelector('svg');
    await expect(svg).toHaveAttribute('viewBox', '0 0 24 24');
    await expect(svg).toHaveAttribute('aria-hidden', 'true');
    await expect(button).toBeInTheDocument();
  },
} satisfies Story;

export const WithLongText = {
  name: 'With Long Text (A1 delvis)',
  args: {
    ...defaultArgs,
    children:
      'Denne knappen har en veldig lang tekst. Så lang at den tvinger fram linjeskift. Her har vi ikke ikon så da skal teksten midtstilles. ' +
      'Denne knappen har en veldig lang tekst. Så lang at den tvinger fram linjeskift. Her har vi ikke ikon så da skal teksten midtstilles',
  },
  argTypes: {
    children: { table: { disable: false } },
  },
} satisfies Story;

export const WithLongTextAndIcon = {
  name: 'With Long Text And Icon (A1 delvis)',
  args: {
    ...defaultArgs,
    svgPath: SendSVGpath,
    children:
      'Denne knappen har en veldig lang tekst. Så lang at den tvinger fram linjeskift. Tekst skal venstrejusteres. ' +
      'Denne knappen har en veldig lang tekst. Så lang at den tvinger fram linjeskift. Tekst skal venstrejusteres.',
  },
  argTypes: {
    children: { table: { disable: false } },
    svgPath: { table: { disable: false } },
  },
} satisfies Story;

export const WithDisabled = {
  name: 'With Disabled (B5)',
  args: {
    ...defaultArgs,
    disabled: true,
  },
  argTypes: {
    disabled: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: {
      focus: `${wrapper} > button`,
      hover: `${wrapper} > button`,
    },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    expect(canvas.getByRole('button')).toBeDisabled();
  },
} satisfies Story;

export const WithDisabledAndIcon = {
  name: 'With Disabled And Icon (B5)',
  args: {
    ...defaultArgs,
    svgPath: SendSVGpath,
    disabled: true,
  },
  argTypes: {
    svgPath: { table: { disable: false } },
    disabled: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: {
      focus: `${wrapper} > button`,
      hover: `${wrapper} > button`,
    },
  },
} satisfies Story;

export const WithType = {
  name: 'With Type (B1)',
  args: {
    ...defaultArgs,
    type: 'submit',
  },
  argTypes: {
    type: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: verifyAttribute('type', 'submit'),
} satisfies Story;

export const WithAriaDescribedby = {
  name: 'With AriaDescribedby (B2)',
  args: {
    ...defaultArgs,
    ariaDescribedby: 'testid1234',
  },
  argTypes: {
    ariaDescribedby: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: verifyAttribute('aria-describedby', 'testid1234'),
} satisfies Story;

export const WithAccesskey = {
  name: 'With Accesskey (B3)',
  args: {
    ...defaultArgs,
    accessKey: 'j',
  },
  argTypes: {
    accessKey: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: verifyAttribute('accessKey', 'j'),
} satisfies Story;

const EventHandlersTemplate: StoryFn<typeof Button> = (args) => {
  const [buttonText, setButtonText] = useState('bruk knapp for å teste events');
  return (
    <Button
      {...args}
      onFocus={(event: FocusEvent<HTMLButtonElement>): void => {
        setButtonText('Knapp har fått fokus');
        args.onFocus && args.onFocus(event);
      }}
      onBlur={(event: FocusEvent<HTMLButtonElement>): void => {
        setButtonText('Knapp har blitt blurret');
        args.onBlur && args.onBlur(event);
      }}
      onClick={(event: MouseEvent<HTMLButtonElement>): void => {
        setButtonText('Knapp har blitt klikket på');
        args.onClick && args.onClick(event);
      }}
    >
      {buttonText}
    </Button>
  );
};

export const WithEventHandlers = {
  render: EventHandlersTemplate,
  name: 'With EventHandlers (A2 delvis)',
  args: {
    ...defaultArgs,
    onFocus: fn(),
    onBlur: fn(),
    onClick: fn(),
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: async ({ args, canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button');
    await expect(button).toBeInTheDocument();
    button.focus();
    await waitFor(() => expect(args.onFocus).toHaveBeenCalled());
    await userEvent.tab();
    await waitFor(() => expect(args.onBlur).toHaveBeenCalled());
    await userEvent.click(button);
    await waitFor(() => expect(args.onClick).toHaveBeenCalled());
  },
} satisfies Story;

const TemplateWithSpinner: StoryFn<typeof Button> = (args) => (
  <>
    <div className={'bottomSpacingXL'}>
      <Button {...args} hasSpinner={false} className={'marginRightM'} />
      <Button
        {...args}
        hasSpinner={false}
        variant={'secondary'}
        className={'marginRightM'}
      />
      <Button
        {...args}
        hasSpinner={false}
        variant={'tertiary'}
        className={'marginRightM'}
      />
      <Button {...args} hasSpinner={false} variant={'danger'} />
    </div>
    <div className={'bottomSpacingXL'}>
      <Button {...args} className={'marginRightM'} />
      <Button {...args} variant={'secondary'} className={'marginRightM'} />
      <Button {...args} variant={'tertiary'} className={'marginRightM'} />
      <Button {...args} variant={'danger'} />
    </div>
    <div className={'bottomSpacingXL'}>
      <Button
        {...args}
        svgPath={SendSVGpath}
        hasSpinner={false}
        className={'marginRightM'}
      />
      <Button
        {...args}
        svgPath={SendSVGpath}
        hasSpinner={false}
        variant={'secondary'}
        className={'marginRightM'}
      />
      <Button
        {...args}
        svgPath={SendSVGpath}
        hasSpinner={false}
        variant={'tertiary'}
        className={'marginRightM'}
      />
      <Button
        {...args}
        svgPath={SendSVGpath}
        hasSpinner={false}
        variant={'danger'}
      />
    </div>
    <Button {...args} svgPath={SendSVGpath} className={'marginRightM'} />
    <Button
      {...args}
      svgPath={SendSVGpath}
      variant={'secondary'}
      className={'marginRightM'}
    />
    <Button
      {...args}
      svgPath={SendSVGpath}
      variant={'tertiary'}
      className={'marginRightM'}
    />
    <Button {...args} svgPath={SendSVGpath} variant={'danger'} />
  </>
);

export const WithSpinner = {
  render: TemplateWithSpinner,
  name: 'With Spinner (A6)',
  args: {
    ...defaultArgs,
    hasSpinner: true,
  },
  argTypes: {
    hasSpinner: { table: { disable: false } },
  },
} satisfies Story;

export const WithExternalIcon = {
  name: 'With External Icon (A7, B6)',
  args: {
    ...defaultArgs,
    isExternal: true,
    children: 'Til altinn.no',
    href: '#',
  },
  argTypes: {
    isExternal: { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const megaButton = canvas.getByRole('button');

    const svg = megaButton.querySelector('svg');
    await expect(svg).toHaveAttribute(
      'aria-label',
      dsI18n.t('ds_buttons:shared.ExternalIcon')
    );
    await expect(svg).toHaveAttribute('viewBox', '0 0 24 24');
    await expect(megaButton).toBeInTheDocument();
  },
} satisfies Story;
