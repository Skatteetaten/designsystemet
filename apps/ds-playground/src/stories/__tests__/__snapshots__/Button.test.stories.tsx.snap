// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tester/Button Defaults play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1   "
          type="button"
  >
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button VariantDanger play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_danger_q67d9_1   "
          type="button"
  >
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button VariantSecondary play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_secondary_q67d9_1   "
          type="button"
  >
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button VariantTertiary play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_tertiary_q67d9_1   "
          type="button"
  >
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithAccesskey play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1   "
          accesskey="j"
          type="button"
  >
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithAriaDescribedby play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1   "
          type="button"
          aria-describedby="testid1234"
  >
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithAttributes play-test 1`] = `
<div data-test-block="true">
  <button id="htmlid"
          class="_button_q67d9_1 _button_primary_q67d9_1   dummyClassname"
          lang="nb"
          data-testid="123ID"
          form="formid123"
          type="button"
  >
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithDisabled play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1   "
          disabled
          type="button"
  >
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithDisabledAndIcon smoke-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1 _button_withIcon_q67d9_47  "
          disabled
          type="button"
  >
    <span class="_iconWrapper_q67d9_25">
      <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
        </path>
      </svg>
    </span>
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithEventHandlers play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1   "
          type="button"
  >
    <span class="_buttonText_q67d9_69">
      Knapp har blitt klikket på
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithExternalIcon play-test 1`] = `
<div data-test-block="true">
  <a class="_button_q67d9_1 _button_primary_q67d9_1  _button_withIconRight_q67d9_51 "
     href="#"
     role="button"
  >
    <span class="_buttonText_q67d9_69">
      Til altinn.no
    </span>
    <span class="_iconWrapper_q67d9_25">
      <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25 _iconExternal_q67d9_66"
           viewbox="0 0 24 24"
           focusable="false"
           aria-label="Til et annet nettsted"
           aria-hidden="false"
           role="img"
      >
        <path d="M14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7Zm5 16H5V5h7V3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7h-2v7Z">
        </path>
      </svg>
    </span>
  </a>
</div>
`;

exports[`Tester/Button WithIcon play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1 _button_withIcon_q67d9_47  "
          type="button"
  >
    <span class="_iconWrapper_q67d9_25">
      <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
        </path>
      </svg>
    </span>
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithLongText smoke-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1   "
          type="button"
  >
    <span class="_buttonText_q67d9_69">
      Denne knappen har en veldig lang tekst. Så lang at den tvinger fram linjeskift. Her har vi ikke ikon så da skal teksten midtstilles. Denne knappen har en veldig lang tekst. Så lang at den tvinger fram linjeskift. Her har vi ikke ikon så da skal teksten midtstilles
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithLongTextAndIcon smoke-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1 _button_withIcon_q67d9_47  "
          type="button"
  >
    <span class="_iconWrapper_q67d9_25">
      <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
        </path>
      </svg>
    </span>
    <span class="_buttonText_q67d9_69">
      Denne knappen har en veldig lang tekst. Så lang at den tvinger fram linjeskift. Tekst skal venstrejusteres. Denne knappen har en veldig lang tekst. Så lang at den tvinger fram linjeskift. Tekst skal venstrejusteres.
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithRef play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1   "
          type="button"
          id="dummyIdForwardedFromRef"
  >
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;

exports[`Tester/Button WithSpinner smoke-test 1`] = `
<div data-test-block="true">
  <div class="bottomSpacingXL">
    <button class="_button_q67d9_1 _button_primary_q67d9_1   marginRightM"
            type="button"
    >
      <span class="_buttonText_q67d9_69">
        Klikk her
      </span>
    </button>
    <button class="_button_q67d9_1 _button_secondary_q67d9_1   marginRightM"
            type="button"
    >
      <span class="_buttonText_q67d9_69">
        Klikk her
      </span>
    </button>
    <button class="_button_q67d9_1 _button_tertiary_q67d9_1   marginRightM"
            type="button"
    >
      <span class="_buttonText_q67d9_69">
        Klikk her
      </span>
    </button>
    <button class="_button_q67d9_1 _button_danger_q67d9_1   "
            type="button"
    >
      <span class="_buttonText_q67d9_69">
        Klikk her
      </span>
    </button>
  </div>
  <div class="bottomSpacingXL">
    <button class="_button_q67d9_1 _button_primary_q67d9_1   marginRightM"
            type="button"
    >
      <span class="_buttonText_q67d9_69 _hide_q67d9_149">
        Klikk her
      </span>
      <div aria-live="polite"
           aria-atomic="true"
           class="_spinner_wnf1f_5 _spinner_white_wnf1f_19 _spinner_flexColumn_wnf1f_32 _spinner_small_wnf1f_22 _spinner_q67d9_153 "
      >
        <div class="_spinnerAnimation_wnf1f_40 ">
        </div>
        <span class="_spinnerTitle_wnf1f_59 _srOnly_wnf1f_70 _spinnerTitle_centerText_wnf1f_66 ">
          Laster inn
        </span>
      </div>
    </button>
    <button class="_button_q67d9_1 _button_secondary_q67d9_1   marginRightM"
            type="button"
    >
      <span class="_buttonText_q67d9_69 _hide_q67d9_149">
        Klikk her
      </span>
      <div aria-live="polite"
           aria-atomic="true"
           class="_spinner_wnf1f_5 _spinner_blue_wnf1f_16 _spinner_flexColumn_wnf1f_32 _spinner_small_wnf1f_22 _spinner_q67d9_153 "
      >
        <div class="_spinnerAnimation_wnf1f_40 ">
        </div>
        <span class="_spinnerTitle_wnf1f_59 _srOnly_wnf1f_70 _spinnerTitle_centerText_wnf1f_66 ">
          Laster inn
        </span>
      </div>
    </button>
    <button class="_button_q67d9_1 _button_tertiary_q67d9_1   marginRightM"
            type="button"
    >
      <span class="_buttonText_q67d9_69 _hide_q67d9_149">
        Klikk her
      </span>
      <div aria-live="polite"
           aria-atomic="true"
           class="_spinner_wnf1f_5 _spinner_blue_wnf1f_16 _spinner_flexColumn_wnf1f_32 _spinner_small_wnf1f_22 _spinner_q67d9_153 "
      >
        <div class="_spinnerAnimation_wnf1f_40 ">
        </div>
        <span class="_spinnerTitle_wnf1f_59 _srOnly_wnf1f_70 _spinnerTitle_centerText_wnf1f_66 ">
          Laster inn
        </span>
      </div>
    </button>
    <button class="_button_q67d9_1 _button_danger_q67d9_1   "
            type="button"
    >
      <span class="_buttonText_q67d9_69 _hide_q67d9_149">
        Klikk her
      </span>
      <div aria-live="polite"
           aria-atomic="true"
           class="_spinner_wnf1f_5 _spinner_black_wnf1f_13 _spinner_flexColumn_wnf1f_32 _spinner_small_wnf1f_22 _spinner_q67d9_153 "
      >
        <div class="_spinnerAnimation_wnf1f_40 ">
        </div>
        <span class="_spinnerTitle_wnf1f_59 _srOnly_wnf1f_70 _spinnerTitle_centerText_wnf1f_66 ">
          Laster inn
        </span>
      </div>
    </button>
  </div>
  <div class="bottomSpacingXL">
    <button class="_button_q67d9_1 _button_primary_q67d9_1 _button_withIcon_q67d9_47  marginRightM"
            type="button"
    >
      <span class="_iconWrapper_q67d9_25">
        <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
             viewbox="0 0 24 24"
             focusable="false"
             aria-hidden="true"
             role="img"
        >
          <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
          </path>
        </svg>
      </span>
      <span class="_buttonText_q67d9_69">
        Klikk her
      </span>
    </button>
    <button class="_button_q67d9_1 _button_secondary_q67d9_1 _button_withIcon_q67d9_47  marginRightM"
            type="button"
    >
      <span class="_iconWrapper_q67d9_25">
        <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
             viewbox="0 0 24 24"
             focusable="false"
             aria-hidden="true"
             role="img"
        >
          <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
          </path>
        </svg>
      </span>
      <span class="_buttonText_q67d9_69">
        Klikk her
      </span>
    </button>
    <button class="_button_q67d9_1 _button_tertiary_q67d9_1 _button_withIcon_q67d9_47  marginRightM"
            type="button"
    >
      <span class="_iconWrapper_q67d9_25">
        <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
             viewbox="0 0 24 24"
             focusable="false"
             aria-hidden="true"
             role="img"
        >
          <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
          </path>
        </svg>
      </span>
      <span class="_buttonText_q67d9_69">
        Klikk her
      </span>
    </button>
    <button class="_button_q67d9_1 _button_danger_q67d9_1 _button_withIcon_q67d9_47  "
            type="button"
    >
      <span class="_iconWrapper_q67d9_25">
        <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
             viewbox="0 0 24 24"
             focusable="false"
             aria-hidden="true"
             role="img"
        >
          <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
          </path>
        </svg>
      </span>
      <span class="_buttonText_q67d9_69">
        Klikk her
      </span>
    </button>
  </div>
  <button class="_button_q67d9_1 _button_primary_q67d9_1 _button_withIcon_q67d9_47  marginRightM"
          type="button"
  >
    <span class="_iconWrapper_q67d9_25 _hide_q67d9_149">
      <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
        </path>
      </svg>
    </span>
    <span class="_buttonText_q67d9_69 _hide_q67d9_149">
      Klikk her
    </span>
    <div aria-live="polite"
         aria-atomic="true"
         class="_spinner_wnf1f_5 _spinner_white_wnf1f_19 _spinner_flexColumn_wnf1f_32 _spinner_small_wnf1f_22 _spinner_q67d9_153 "
    >
      <div class="_spinnerAnimation_wnf1f_40 ">
      </div>
      <span class="_spinnerTitle_wnf1f_59 _srOnly_wnf1f_70 _spinnerTitle_centerText_wnf1f_66 ">
        Laster inn
      </span>
    </div>
  </button>
  <button class="_button_q67d9_1 _button_secondary_q67d9_1 _button_withIcon_q67d9_47  marginRightM"
          type="button"
  >
    <span class="_iconWrapper_q67d9_25 _hide_q67d9_149">
      <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
        </path>
      </svg>
    </span>
    <span class="_buttonText_q67d9_69 _hide_q67d9_149">
      Klikk her
    </span>
    <div aria-live="polite"
         aria-atomic="true"
         class="_spinner_wnf1f_5 _spinner_blue_wnf1f_16 _spinner_flexColumn_wnf1f_32 _spinner_small_wnf1f_22 _spinner_q67d9_153 "
    >
      <div class="_spinnerAnimation_wnf1f_40 ">
      </div>
      <span class="_spinnerTitle_wnf1f_59 _srOnly_wnf1f_70 _spinnerTitle_centerText_wnf1f_66 ">
        Laster inn
      </span>
    </div>
  </button>
  <button class="_button_q67d9_1 _button_tertiary_q67d9_1 _button_withIcon_q67d9_47  marginRightM"
          type="button"
  >
    <span class="_iconWrapper_q67d9_25 _hide_q67d9_149">
      <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
        </path>
      </svg>
    </span>
    <span class="_buttonText_q67d9_69 _hide_q67d9_149">
      Klikk her
    </span>
    <div aria-live="polite"
         aria-atomic="true"
         class="_spinner_wnf1f_5 _spinner_blue_wnf1f_16 _spinner_flexColumn_wnf1f_32 _spinner_small_wnf1f_22 _spinner_q67d9_153 "
    >
      <div class="_spinnerAnimation_wnf1f_40 ">
      </div>
      <span class="_spinnerTitle_wnf1f_59 _srOnly_wnf1f_70 _spinnerTitle_centerText_wnf1f_66 ">
        Laster inn
      </span>
    </div>
  </button>
  <button class="_button_q67d9_1 _button_danger_q67d9_1 _button_withIcon_q67d9_47  "
          type="button"
  >
    <span class="_iconWrapper_q67d9_25 _hide_q67d9_149">
      <svg class="_systemIcon_6zbcy_2 _systemIcon_medium_6zbcy_9 _icon_q67d9_25"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M2 3v7l15 2-15 2v7l21-9L2 3Z">
        </path>
      </svg>
    </span>
    <span class="_buttonText_q67d9_69 _hide_q67d9_149">
      Klikk her
    </span>
    <div aria-live="polite"
         aria-atomic="true"
         class="_spinner_wnf1f_5 _spinner_black_wnf1f_13 _spinner_flexColumn_wnf1f_32 _spinner_small_wnf1f_22 _spinner_q67d9_153 "
    >
      <div class="_spinnerAnimation_wnf1f_40 ">
      </div>
      <span class="_spinnerTitle_wnf1f_59 _srOnly_wnf1f_70 _spinnerTitle_centerText_wnf1f_66 ">
        Laster inn
      </span>
    </div>
  </button>
</div>
`;

exports[`Tester/Button WithType play-test 1`] = `
<div data-test-block="true">
  <button class="_button_q67d9_1 _button_primary_q67d9_1   "
          type="submit"
  >
    <span class="_buttonText_q67d9_69">
      Klikk her
    </span>
  </button>
</div>
`;
